SQL-733: Implemented custom Tableau Connector files

bucaojit opened the pull request:
JIRA: https://jira.mongodb.org/browse/SQL-733

Implementing the files other than the Capabilities section of the `manifest.xml` and Dialect definition file `dialect.tdd`.

bucaojit committed:
SQL-733: Implement custom Tableau Connector files

bucaojit committed:
Merge branch 'master' into SQL-733

bucaojit commented on connector/connectionFields.xml:
```diff
@@ -1,6 +1,14 @@
 <?xml version="1.0" encoding="utf-8"?>
 
 <connection-fields>
+
```
Screenshot of the connection dialogue [here](https://docs.google.com/document/d/1vt6f0c2T74jgb_QqLFM8FEM12dh56BfD2Qh5Gw6rxWc/edit#bookmark=id.oc2z3uwg2ulz)

bucaojit commented on connector/connectionResolver.tdr:
```diff
@@ -1,24 +1,11 @@
-    <connection-builder>
-      <script file='connectionBuilder.js'/>
-    </connection-builder>
-    <connection-normalizer>
```
Removing connection-normalizer to utilize inferred connection resolver.

bucaojit replied:
Re-added to work with older version 2020.4

rychipman commented on connector/connectionFields.xml:
```diff
@@ -1,6 +1,14 @@
 
 <connection-fields>
+
+  <field name="v-prefix" label="Connection Prefix" value-type="selection" category="endpoint" >
```
It seems like "protocol" might be a more appropriate name, since that's what `mongodb` and `mongodb+srv` are

bucaojit replied:
Removed this field as we are using connection URI

rychipman commented on connector/connectionFields.xml:
```diff
@@ -11,6 +19,29 @@
+    <conditions>
+      <condition field="v-logdiroption" value="required"/>
+    </conditions>
+  </field>
```
```suggestion
```
suggestion: can we leave this out for now and talk further in the design doc about whether we want it? This seems like it might be a bit too niche to merit inclusion in the UI itself, but I think we should defer it for further discussion

rychipman commented on connector/connectionFields.xml:
```diff
@@ -11,6 +19,29 @@
 
   <field name="password" label="Password" value-type="string" category="authentication" secure="true" />
 
+  <field name="v-authDatabase" label="Authentication Database" value-type="string" category="authentication" default-value="admin" />
```
I think we should talk about the general stragegy here wrt various MongoDB URI parameters. Specifically, the way this patch is going seems to suggest that any URI parameter we want to enable will require a separate configuration UI element to be added.

I'm curious to know whether/why this is preferable to just allowing the user to provide the whole MongoDB URI

nbagnard replied:
I share some of the reserve Ryan has here. It would be nice to be able to let the user also input his custom jdbc url, because otherwise we are restricting them to only the parameters that are in the dialog. These here are the ones we use, but other users can have more refined urls best suited for their needs. 
The issue I see is that not all the parameters are supported in the jdbc url. dialect isn't (but it's not visible for Tableau users, so not a blocker), logging params won't be though unless we add support for them in the jdbc driver itself. 
In short, if not all parameters are supported in the url, we can't rely on only letting the user provide its url. Maybe a conditional display which only "asks" for the parameters which are not supported in the url when opting for the non default url would do the trick? 
I'm also (still) very much in-favour of supporting all parameters as part of the jdbc url ;)

bucaojit replied:
Spoke with Ryan offline and as the current iteration has few properties to set such as 'database', I will make the change to allow the user to provide the mongodb URI and for the additional properties we can discuss the design of when we get to it.

rychipman commented on connector/manifest.xml:
```diff
@@ -1,14 +1,14 @@
 <?xml version='1.0' encoding='utf-8' ?>
 
-<connector-plugin class='mongosql_jdbc' superclass='jdbc' plugin-version='0.0.0' name='MongoSQL' version='18.1' min-version-tableau='2020.4'>
+<connector-plugin class='mongosql_jdbc' superclass='jdbc' plugin-version='0.1.0' name='MongoSQL' version='18.1' min-version-tableau='2021.1'>
```
Let's talk with Alexi about this, but I don't _think_ we want to use "MongoSQL" here. Instead, I think "Atlas SQL" or "MongoDB" will be preferred, depending on the context

nbagnard replied:
Are we planning to also provide a connector bundle for ODBC at some point?
If so, having a JDBC suffix or prefix would make it easier for us in the future to add the ODBC connector and avoid confusion.
Also as a side note, I personally appreciate when a connector tells me specifically which kind of driver it is using underneath. It might not be as useful for non technical users, but the ones familiar with ODBC and JDBC will likely be happy to know as well.

bucaojit replied:
For Tableau there aren't plans to provide a connector for ODBC.  
In the "Installed Connectors" section the connector names are generally the database name.  I do see some connectors in the "Additional Connectors" that show JDBC or ODBC, I'm not sure yet why there are the two sections.

nbagnard replied:
Sounds good.
I guess if there is only flavour of the connector then it does make sense to not have JDBC/ODBC.
When I say "at some point", I looking at a few years from now. If we know that we will never want to provide a connector using the ODBC driver, that's fine. And if we do, then we'll find a differentiator at this time :)

rychipman commented on connector/manifest.xml:
```diff
@@ -1,14 +1,14 @@
-      <support-link url="https://example.com"/>
-      <driver-download-link url="https://drivers.example.com"/>
+      <support-link url="https://support.mongodb.com"/>
+      <driver-download-link url="https://search.maven.org/search?q=a:mongodb-jdbc"/>
```
```suggestion
      <driver-download-link url="https://search.maven.org/artifact/org.mongodb/mongodb-jdbc"/>
```

rychipman replied:
Another thought/question: where does this download link appear in the connector? Is it intended for users to follow and download manually, or is it used automatically by Tableau somehow?

bucaojit replied:
I see this download link show up in the error dialogue when the jdbc driver isn't present in the Tableau/Drivers/ directory.  It's meant for the user to follow and install manually.  With the following text:

No suitable driver installed, or the URL is incorrect
Connector Class: mongosql_jdbc, Version: 0.1.0
[Download and install the drivers, and then connect.](https://search.maven.org/artifact/org.mongodb/mongodb-jdbc)
[For support, contact MongoDB.](https://support.mongodb.com/)

rychipman commented on connector/connectionProperties.js:
```diff
@@ -2,9 +2,20 @@
     props["dialect"] = "mongosql";
-    props["database"] = attr[connectionHelper.attributeDatabase]
-    props["authSource"] = "admin";
+    props["TYPE_MAPPING_MODE"] = "tableau";
```
```suggestion
```
I don't think this property exists in the JDBD driver, does it?

bucaojit replied:
It is in the design for the type mapping, I will remove it for now.

rychipman commented on connector/manifest.xml:
```diff
@@ -1,14 +1,14 @@
   <connection-customization class="mongosql_jdbc" enabled="true" version="10.0">
-    <vendor name="vendor"/>
-    <driver name="driver"/>
+    <vendor name="MongoDB"/>
+    <driver name="MongoSQL Connector"/>
```
do you know where these values actually get used by the connector? I couldn't find any explanation in the docs

bucaojit replied:
I don't see where it is used either, it doesn't show up in the logs.  The fields are needed to pass the packager validation however.

nbagnard replied:
Weird, their doc says "Vendor defined attributes will be logged and persisted to Tableau workbook xml in plain text." https://tableau.github.io/connector-plugin-sdk/docs/connection-dialog
It also seems that they can be used as connection attributes, but I don't think we need them, do we?

bucaojit replied:
The vendor defined attributes referred to there are the ones defined in the connectionFields.xml that start with the `v-` like the `v-loglevel` for example.
The fields in connectionFields.xml are the ones that would be used for the `<required-attributes>` element, which we wouldn't need if we use the inferred connection resolver, added in the 2021.1 release.

nbagnard replied:
I see.
Well, then it will stay a mystery. It could also be something related to a Tableau server user case. Who knows...

rychipman commented on connector/connectionMetadata.xml:
```diff
@@ -1,8 +1,9 @@
-  <database enabled='true'>
-    <field />
+  <database enabled='true' label='Database'>
+    <field optional='true' />
```
What happens if the user doesn't provide a value for the database? Don't many of the JDBC driver's API calls require a database to be set in order to work?

bucaojit replied:
That's a good point.  If the field isn't set, I see Tableau is able to connect but the user needs to choose the database from the drop-down menu.  In the logs I see Tableau making a  `MongoDatabaseMetaData` call to get the database names.  
I can change the database field to being required.

rychipman reviewed (changes requested):
Great start! some small suggestions and a bunch of questions from me

nbagnard commented on connector/connectionFields.xml:
```diff
@@ -11,6 +19,29 @@
 
+  <field name="v-authDatabase" label="Authentication Database" value-type="string" category="authentication" default-value="admin" />
+
+  <field name="v-loglevel" label="JDBC Log Level" value-type="selection" default-value="log-info" category="general" >
```
Shall we wait for logging to be implemented and available before adding this to the dialog? It seems a bit odd to have it in the dialog before it actually is available.

bucaojit replied:
Sure, it's removed.

nbagnard commented on connector/manifest.xml:
```diff
@@ -1,14 +1,14 @@
 <?xml version='1.0' encoding='utf-8' ?>
 
-<connector-plugin class='mongosql_jdbc' superclass='jdbc' plugin-version='0.0.0' name='MongoSQL' version='18.1' min-version-tableau='2020.4'>
+<connector-plugin class='mongosql_jdbc' superclass='jdbc' plugin-version='0.1.0' name='MongoSQL' version='18.1' min-version-tableau='2021.1'>
```
Is _min-version-tableau='2021.1'_ not going to be too restrictive and cut us from potential users who are using a slightly old version of Tableau?
Was _2020.4_ looking to old?

bucaojit replied:
Good question, the 2021.1 version allows for the inferred connection resolver, which is used in the connector however may not be necessary.  I'll test out with setting the resolver and checking again if there is a benefit.

bucaojit replied:
I added back the `connection-normalizer` and tested with 2020.4 and it worked fine.  2020.3 is the first version that uses the Connection Dialogs V2, but that version is no longer available due to the log4j vulnerability so `2020.4` is a good min version.

bucaojit committed:
Update connector/connectionFields.xml

Co-authored-by: Ryan Chipman <git@mail.ryanchipman.com>

bucaojit committed:
Update connector/manifest.xml

Co-authored-by: Ryan Chipman <git@mail.ryanchipman.com>

bucaojit committed:
Update connector/connectionProperties.js

Co-authored-by: Ryan Chipman <git@mail.ryanchipman.com>

bucaojit committed:
Addressing review comments

bucaojit committed:
Added connection-normalizer, removed logging properties

bucaojit commented:
Updated the connection [dialogue screenshot](https://docs.google.com/document/d/1vt6f0c2T74jgb_QqLFM8FEM12dh56BfD2Qh5Gw6rxWc/edit#bookmark=id.oc2z3uwg2ulz)

Checked with Alexi on the naming, updated name to "MongoDB Atlas".

rychipman commented on connector/connectionFields.xml:
```diff
@@ -1,9 +1,8 @@
-  <field name="server" label="Server" value-type="string" category="endpoint" />
 
-  <field name="port" label="Port" value-type="string" category="endpoint" default-value="27017" />
+  <field name="v-connection-uri" label="Connection URI" value-type="string" category="endpoint" />
```
```suggestion
  <field name="v-mongodb-uri" label="MongoDB URI" value-type="string" category="endpoint" />
```
nit: let's change this name to make it very clear that this is a MongoDB URI, not some other URI

rychipman commented on connector/manifest.xml:
```diff
@@ -1,14 +1,14 @@
 <?xml version='1.0' encoding='utf-8' ?>
 
-<connector-plugin class='mongosql_jdbc' superclass='jdbc' plugin-version='0.0.0' name='MongoSQL' version='18.1' min-version-tableau='2020.4'>
+<connector-plugin class='mongosql_jdbc' superclass='jdbc' plugin-version='0.1.0' name='MongoDB Atlas' version='18.1' min-version-tableau='2020.4'>
```
```suggestion
<connector-plugin class='mongodb_jdbc' superclass='jdbc' plugin-version='0.1.0' name='MongoDB Atlas' version='18.1' min-version-tableau='2020.4'>
```
let's change this "mongosql" to "mongodb" as well

rychipman commented on connector/connectionProperties.js:
```diff
@@ -2,9 +2,8 @@
-    props["authSource"] = "admin";
 
     if (attr[connectionHelper.attributeSSLMode] == "require") {
         props["ssl"] = "true";
```
Are SSL properties currently handled at all by the JDBC driver? I don't see any tests asserting that we handle them in `MongoDriverTest.java`. Beyond that, I think we also have additional questions to answer around how to handle conflicts between SSL options in the properties and in the URI.

If there's not a clear and obvious answer to these things at the moment, I would suggest that we remove this for now and create a new design doc section to discuss, so that this question doesn't hold up merging the rest of the PR

bucaojit replied:
I see that the `ssl` property gets added to the ConnectionString in the driver when I print the connection string.  I have also noticed that whether that property is set or not affects whether the driver can connect to ADL, with the local ADL instance it only works if unset.

I will remove the SSL UI element and users can specify it in the MongoDB URI

rychipman reviewed (changes requested):
Just a few small things remaining, otherwise looks great!

bucaojit committed:
Removing sslmode property

bucaojit committed:
Update connector/manifest.xml

Co-authored-by: Ryan Chipman <git@mail.ryanchipman.com>

bucaojit committed:
Update connector/connectionFields.xml

Co-authored-by: Ryan Chipman <git@mail.ryanchipman.com>

bucaojit committed:
Removing SSL sections

bucaojit committed:
Naming change

bucaojit committed:
Updating name in evergreen file

bucaojit committed:
Removing sslmode from attribute list

rychipman reviewed (approved):
LGTM!

nbagnard reviewed (approved):
LGTM!

bucaojit commented:
evergreen merge
